R/comment_linters.R

```diff
@@ -0,0 +1,169 @@
+ops <- list(
+  "+",
+  #"-",
+  "=",
+  "==",
+  "!=",
+  "<=",
+  ">=",
+  "<-",
+  "<<-",
+  "<",
+  ">",
+  "->",
+  "->>",
+  "%%",
+  "/",
+  "^",
+  "*",
+  "**",
+  "|",
+  "||",
+  "&",
+  "&&",
+  rex("%", except_any_of("%"), "%"))
+
+#' @describeIn linters Check that there is no commented code outside roxygen
+#' blocks
+#' @export
+commented_code_linter <- function(source_file) {
+  res <- re_matches(source_file$file_lines,
+                    rex("#", any_spaces,
+                        capture(name = "code",
+                          # except("'"),
+                          anything,
+                          or(some_of("{}[]"), # code-like parentheses
+                            or(ops), # any operator
+                            group(graphs, "(", anything, ")"), # a function call
+                            group("!", alphas) # a negation
+                            ),
+                          anything
+                        )
+                    ),
+                    global = FALSE, locations = TRUE)
+
+  line_numbers <- rownames(na.omit(res))
+  lapply(line_numbers, function(line_number) {
+    line <- source_file$file_lines[as.numeric(line_number)]
+    is_parsable <- parsable(substr(line,
+                                   res[line_number, "code.start"],
+                                   res[line_number, "code.end"]))
+    if (is_parsable) {
+      Lint(
+        filename = source_file$filename,
+        line_number = line_number,
+        column_number = res[line_number, "code.start"],
+        type = "style",
+        message = "Commented code should be removed.",
+        line = line,
+        linter = "commented_code_linter",
+        ranges = list(c(res[line_number, "code.start"], res[line_number, "code.end"]))
+        )
+    }
+  })
+}
+
+# is given text parsable
+parsable <- function(x) {
+  if (is.null(x)) {
+    return(FALSE)
+  }
+  res <- try(parse(text = x), silent = TRUE)
+
+  !inherits(res, "try-error")
+}
+
+
+
+#' @describeIn linters  Check that the source contains no TODO comments (case-insensitive).
+#' @param todo  Vector of strings that identify TODO comments.
+#' @export
+todo_comment_linter <- function(todo=c("todo", "fixme")) {
```

Comment:
The function is more complicated than anticipated because not all comments are reported within `source_file$parsed_content`.

```R
# TODO 1
x <- 3 # TODO 2
foo <- function(x) { # TODO 3
  # TODO 4
  x * 2 # TODO 5
}
y <- foo(x) # TODO 6
# TODO 7
```
In the example above, only TODO 3-5 are returned as tokens. In `semicolon_terminator_linter()`, there is a similar issue that is addressed by the `locate_other_sc` function.